File: sort/merge_sort/python/merge_sort2.py
print("Enter the list of numbers to be sorted:")
s=input()
if(s[-1]==" "):
	s=s[0:len(s)-1:1]
s=list(map(int,s.split(" ")))

#MERGE SORT
def merge(a,b,c):
	k=0
	i=0
	j=0
	while(len(a)>i and len(b)>j):
		if(a[i]>=b[j]):
			c[k]=b[j]
			j+=1
		else:
			c[k]=a[i]
			i+=1
		k+=1
	while(len(a)>i):
		c[k]=a[i]
		k+=1
		i+=1
	while(len(b)>j):
		c[k]=b[j]
		k+=1
		j+=1
def mergesort(l):
	if(len(l)<2):
		return
	mid=len(l)//2
	L=l[0:mid:1]
	R=l[mid:len(l):1]
	mergesort(L)
	mergesort(R)
	merge(L,R,l)
mergesort(s)
print(" ".join(list(map(str,s))))

File: sort/merge_sort/c++/mergesort2.cpp
/* Kunal Garg
   IIIT Hyderabad  */
#include<iostream>
#include<cstdio>
#include<string>
#include<algorithm>
#include <vector>
int arr[1111111];
int temp[1111111];
int n;
using namespace std;

void merge(int l,int m,int r)
{
	int mid = m+1;
	int k=1;
	int safe = l;
	while(l<=m && mid<=r)
	{ 
		if(arr[l] <= arr[mid])
			temp[k++] = arr[l++];
		else 
			temp[k++] = arr[mid++];
	}
	while(l<=m)
		temp[k++] = arr[l++];
	while(mid <=r)
		temp[k++] = arr[mid++];
	for(int i=1;i<k;i++)
		arr[safe+i-1] = temp[i];
	return;
}

void merge_sort(int l,int r)
{
	if(l<r)
	{
		int m = (l+r)>>1;
		merge_sort(l,m);
		merge_sort(m+1,r);
		merge(l,m,r);
		return;
	}
	return;
}

int main()
{

	int pos=1;
	int x;
	int m;
	cin >> m;
	for(int i=1;i<=m;i++)
		cin >> arr[i];
	merge_sort(1,m);
	for(int i=1;i<=m;i++)
		printf("%d ",arr[i]);
	printf("\n");
}





